Milvus.Client/MilvusCollection.Entity.cs

```diff
@@ -418,61 +418,144 @@ public Task<FlushResult> FlushAsync(CancellationToken cancellationToken = defaul
+    public async IAsyncEnumerable<IReadOnlyList<FieldData>> QueryWithIteratorAsync(
+        string? expression = null,
+        int batchSize = 1000,
+        QueryParameters? parameters = null,
+        [EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        if ((parameters?.Offset ?? 0) != 0)
         {
-            if (parameters.TimeTravelTimestamp is not null)
+            throw new MilvusException("Not support offset when searching iteration");
+        }
+
+        var describeResponse = await _client.InvokeAsync(
+                _client.GrpcClient.DescribeCollectionAsync,
+                new DescribeCollectionRequest { CollectionName = Name },
+                r => r.Status,
+                cancellationToken)
+                .ConfigureAwait(false);
+
+        var pkField = describeResponse.Schema.Fields.FirstOrDefault(x => x.IsPrimaryKey);
+        if (pkField == null)
+        {
+            throw new MilvusException("Schema must contain pk field");
+        }
+
+        var isUserRequestPkField = parameters?.OutputFieldsInternal?.Contains(pkField.Name) ?? false;
+        var userExpression = expression;
+        var userLimit = parameters?.Limit ?? int.MaxValue;
+
+        QueryRequest request = new()
+        {
+            CollectionName = Name,
+            Expr = (userExpression, pkField) switch
             {
-                request.TravelTimestamp = parameters.TimeTravelTimestamp.Value;
+                // If user expression is not null, we should use it
+                {userExpression: not null} => userExpression,
+                // If user expression is null and pk field is string
+                {pkField.DataType: DataType.VarChar} => $"{pkField.Name} != ''",
+                // If user expression is null and pk field is not string
+                _ => $"{pkField.Name} < {long.MaxValue}",
             }
+        };
 
-            if (parameters.PartitionNamesInternal?.Count > 0)
+        PopulateQueryRequestFromParameters(request, parameters);
+
+        // Request id field in any case to proceed with an iterations
+        if (!isUserRequestPkField) request.OutputFields.Add(pkField.Name);
+
+        // Replace parameters required for iterator
+        ReplaceKeyValueItems(request.QueryParams,
+            new Grpc.KeyValuePair {Key = Constants.Iterator, Value = "True"},
+            new Grpc.KeyValuePair {Key = Constants.ReduceStopForBest, Value = "True"},
+            new Grpc.KeyValuePair {Key = Constants.BatchSize, Value = batchSize.ToString(CultureInfo.InvariantCulture)},
+            new Grpc.KeyValuePair {Key = Constants.Offset, Value = 0.ToString(CultureInfo.InvariantCulture)},
+            new Grpc.KeyValuePair {Key = Constants.Limit, Value = Math.Min(batchSize, userLimit).ToString(CultureInfo.InvariantCulture)});
+
+        var processedItemsCount = 0;
+        while (!cancellationToken.IsCancellationRequested)
```

Comment:
Because then it will be while(true) so it is defensive strategy so I will be sure that loop will broke.

I prefer to leave this. What do you think?

Reply:
Adding the check doesn't ensure that the loop will be broken any more than a simple `while (true)`. The thing that guarantees that the loop will break is the `yield break` inside, when there are no more results. And since the cancellation token is checked within InvokeAsync anyway, I think this should be removed.

Reply:
Changed